cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageStore.java

```diff
@@ -0,0 +1,150 @@
+   * @param datasetInstance dataset accessed by the program
+   * @param accessType access type
+   * @param metadata metadata to store for the access
+   * @param component program component such as flowlet id, etc.
+   */
+  public void addAccess(final Id.Run run, final Id.DatasetInstance datasetInstance,
+                        final AccessType accessType, final String metadata, @Nullable final Id component) {
+    execute(new TransactionExecutor.Procedure<LineageDataset>() {
+      @Override
+      public void apply(LineageDataset input) throws Exception {
+        input.addAccess(run, datasetInstance, accessType, metadata, component);
+      }
+    });
+  }
+
+  /**
+   * Fetch program-dataset access information for a dataset for a given period.
+   *
+   * @param datasetInstance dataset for which to fetch access information
+   * @param start start time period
+   * @param end end time period
+   * @return program-dataset access information
+   */
+  public Set<Relation> getRelations(final Id.DatasetInstance datasetInstance, final long start, final long end) {
+    return execute(new TransactionExecutor.Function<LineageDataset, Set<Relation>>() {
+      @Override
+      public Set<Relation> apply(LineageDataset input) throws Exception {
+        return input.getRelations(datasetInstance, start, end);
+      }
+    });
+  }
+
+  /**
+   * Fetch program-dataset access information for a program for a given period.
+   *
+   * @param program program for which to fetch access information
+   * @param start start time period
+   * @param end end time period
+   * @return program-dataset access information
+   */
+  public Set<Relation> getRelations(final Id.Program program, final long start, final long end) {
+    return execute(new TransactionExecutor.Function<LineageDataset, Set<Relation>>() {
+      @Override
+      public Set<Relation> apply(LineageDataset input) throws Exception {
+        return input.getRelations(program, start, end);
+      }
+    });
+  }
+
+  private <T> T execute(TransactionExecutor.Function<LineageDataset, T> func) {
+    try {
+      LineageDataset lineageDataset = getLineageDataset();
+      TransactionExecutor txExecutor = Transactions.createTransactionExecutor(executorFactory, lineageDataset);
+      return txExecutor.execute(func, lineageDataset);
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error accessing %s dataset", LINEAGE_DATASET_ID), e);
+    }
+  }
+
+  private void execute(TransactionExecutor.Procedure<LineageDataset> func) {
+    try {
+      LineageDataset lineageDataset = getLineageDataset();
+      TransactionExecutor txExecutor = Transactions.createTransactionExecutor(executorFactory, lineageDataset);
+      txExecutor.execute(func, lineageDataset);
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error accessing %s dataset", LINEAGE_DATASET_ID), e);
+    }
+  }
+
+  private LineageDataset getLineageDataset()
```

Comment:
To make it clear that it creates a new instance on every call (which is needed because dataset is not thread safe), better call this method `newLineageDataset`.